src/org/jetbrains/plugins/scala/lang/psi/types/ScSubstitutor.scala

```diff
@@ -8,20 +8,22 @@ import com.intellij.psi._
 import org.jetbrains.plugins.scala.lang.psi.api.base.ScFieldId
 import org.jetbrains.plugins.scala.lang.psi.api.base.patterns.ScBindingPattern
 import org.jetbrains.plugins.scala.lang.psi.api.statements.ScFunction
-import org.jetbrains.plugins.scala.lang.psi.api.statements.params.{ScTypeParam, ScParameter}
+import org.jetbrains.plugins.scala.lang.psi.api.statements.params.ScParameter
 import org.jetbrains.plugins.scala.lang.psi.api.toplevel.ScTypedDefinition
 import org.jetbrains.plugins.scala.lang.psi.api.toplevel.typedef.{ScTemplateDefinition, ScTypeDefinition}
 import org.jetbrains.plugins.scala.lang.psi.types.nonvalue.{Parameter, ScMethodType, ScTypePolymorphicType, TypeParameter}
 import org.jetbrains.plugins.scala.lang.psi.types.result.TypingContext
 
-import scala.annotation.tailrec
 import scala.collection.immutable.{HashMap, HashSet, Map}
 
 /**
 * @author ven
 */
 object ScSubstitutor {
-  val empty: ScSubstitutor = new ScSubstitutor()
+  val empty: ScSubstitutor = EmptySubstitutor
```

Comment:
@Alefas Actually, what you are suggesting will not work, a lot of code just won't compile in the places where ScSubstitutor type is inferred from ScSubstitutor.empty. Here is an example:
````
def foo() = {
  Seq(1, 2, 3).foldLeft(ScSubstitutor.empty) {(s, p) =>
    s.bindT(null, null) 
    //Type mismatch: expression of type ScSubstitutor does not conform to expected type ScSubstitutor.empty.type
  }
}
````